src/core/src/dimension_tracker.cpp

```diff
@@ -8,26 +8,23 @@ using namespace ov;
 
 void TableOfEquivalence::set_as_equal(const Dimension& lhs, const Dimension& rhs) {
     const auto &l_label = DimensionTracker::get_label(lhs), r_label = DimensionTracker::get_label(rhs);
-    bool l_known = dimension_table_of_equivalence.count(l_label) && dimension_table_of_equivalence[l_label],
-         r_known = dimension_table_of_equivalence.count(r_label) && dimension_table_of_equivalence[r_label];
-    if (l_known && r_known) {
-        auto soup_l = dimension_table_of_equivalence[l_label];
-        soup_l->insert(r_label);
-        auto soup_r = dimension_table_of_equivalence[r_label];
-        soup_r->insert(l_label);
-        soup_l->insert(soup_r->begin(), soup_r->end());
-        soup_r->insert(soup_l->begin(), soup_l->end());
-    } else {
-        auto soup = std::make_shared<std::set<label_t>>();
-        if (l_known)
-            soup = dimension_table_of_equivalence[l_label];
-        else if (r_known)
-            soup = dimension_table_of_equivalence[r_label];
-        soup->insert(l_label);
-        soup->insert(r_label);
-        dimension_table_of_equivalence[l_label] = soup;
-        dimension_table_of_equivalence[r_label] = soup;
-    }
+    if (l_label == ov::no_label || r_label == ov::no_label)
+        // TODO after value restriction enabling: non labeled dim propagates restriction (if any) to labeled dim
+        return;
+
+    auto get_soup = [](const label_t& label, EqTable& table) -> EqualitySoup {
+        if (!table.count(label) || !table.at(label))
+            table[label] = std::make_shared<std::set<label_t>>(std::set<label_t>{label});
+        return table.at(label);
```

Comment:
```suggestion
        auto res = table[label];
        if (!res)
            res = std::make_shared<std::set<label_t>>();
        return res;
```
checked:
![image](https://github.com/openvinotoolkit/openvino/assets/21181928/1e283a58-83b7-44fb-9ac8-6cb7174b3c18)

Reply:
I don't like the idea of adding new records pre each operator[] call -- seems excessive.

Reply:
Another approach can be to use `find` if not end then return it, otherwise use `emplace` and emplace return iterator to created element.

Reply:
But you add it anyway
![image](https://github.com/openvinotoolkit/openvino/assets/21181928/a3b265c5-9f38-41f8-abdf-72e91586a2d9)
if the label doesn't exist in the table, it will be created.
I believe the suggested code above works the same

Reply:
`find` will be better